doc/examples/developers/plot_threshold_li.py

```diff
@@ -0,0 +1,233 @@
+ax[2].plot(iter_thresholds3, iter_entropies3, label='optimization path')
+ax[2].scatter(iter_thresholds3, iter_entropies3, c='C1')
+ax[2].legend(loc='upper right')
+
+plt.show()
+
+###############################################################################
+# To see what is going on, let's define a function, ``li_gradient``, that
+# replicates the inner loop of the Li method and returns the *change* from the
+# current threshold value to the next one. When this gradient is 0, we are at
+# a so-called *stationary point* and Li returns this value. When it is
+# negative, the next threshold guess will be lower, and when it is positive,
+# the next guess will be higher.
+#
+# In the plot below, we show the cross-entropy and the Li update path when the
+# initial guess is on the *right* side of that entropy peak. We overlay the
+# threshold update gradient, marking the 0 gradient line and the default
+# initial guess by ``threshold_li``.
+
+def li_gradient(image, t):
+    """Find the threshold update at a given threshold."""
+    foreground = image > t
+    mean_fore = np.mean(image[foreground])
+    mean_back = np.mean(image[~foreground])
+    t_next = ((mean_back - mean_fore) /
+              (np.log(mean_back) - np.log(mean_fore)))
+    dt = t_next - t
+    return dt
+
+
+iter_thresholds4 = []
+opt_threshold4 = filters.threshold_li(cell, initial_guess=68,
+                                      iter_callback=iter_thresholds4.append)
+iter_entropies4 = [_cross_entropy(cell, t) for t in iter_thresholds4]
+print(len(iter_thresholds4), 'examined, optimum:', opt_threshold4)
+
+gradients = [li_gradient(cell, t) for t in thresholds2]
+
+fig, ax1 = plt.subplots()
+ax1.plot(thresholds2, entropies2)
+ax1.plot(iter_thresholds4, iter_entropies4)
+ax1.scatter(iter_thresholds4, iter_entropies4, c='C1')
+ax1.set_xlabel('threshold')
+ax1.set_ylabel('cross entropy', color='C0')
+ax1.tick_params(axis='y', labelcolor='C0')
+
+ax2 = ax1.twinx()
+ax2.plot(thresholds2, gradients, c='C3')
+ax2.hlines([0], xmin=thresholds2[0], xmax=thresholds2[-1],
+           colors='gray', linestyles='dashed')
+ax2.vlines(np.mean(cell), ymin=np.min(gradients), ymax=np.max(gradients),
+           colors='gray', linestyles='dashed')
+ax2.set_ylabel('$\Delta$(threshold)', color='C3')
+ax2.tick_params(axis='y', labelcolor='C3')
+
+fig.tight_layout()
+
+plt.show()
+
+###############################################################################
+# In addition to allowing users to provide a number as an initial guess,
+# :func:`skimage.filters.threshold_li` can receive a function that makes a
+# guess from the image intensities, just like :func:`numpy.mean` does by
+# default. This might be a good option when many images with different ranges
+# need to be processed.
+
+def quantile_95(image):
+    return np.quantile(image, 0.95)
+
+iter_thresholds5 = []
+opt_threshold5 = filters.threshold_li(cell, initial_guess=quantile_95,
```

Comment:
```suggestion
opt_threshold5 = filters.threshold_li(cell, initial_guess=np.quantile(cell, 0.95),
```

Following my other comment, this acheives the same thing, and really reduces the scope of the API.

Users that want to make things more generic, will know how to make their own function wrapper.

Reply:
@hmaarrfk this is fine for one image, but dramatically increases the code complexity for a large image collection.

Reply:
(and doesn't solve the problem that we do a lot of sanitising of the image, which the user won't benefit from here)

Reply:
Ah. Ok. The sanitization point makes sense.